lib/bamboo/mailer.ex

```diff
@@ -132,51 +168,91 @@ defmodule Bamboo.Mailer do
   with `deliver_later/1`.
+
+  If the email is successfully scheduled for delivery, this function will return
+  an `{:ok, email}`.
+
+  If the email is invalid, this function will return an `{:error, error}` tuple.
   """
   def deliver_later(_email, _opts \\ []) do
     raise @cannot_call_directly_error
   end
 
-  @doc false
-  def deliver_now(adapter, email, config, response: true) do
-    email = email |> validate_and_normalize(adapter)
+  @doc """
+  Deliver an email in the background.
 
-    if email.to == [] && email.cc == [] && email.bcc == [] do
-      debug_unsent(email)
-      email
-    else
-      debug_sent(email, adapter)
-      response = adapter.deliver(email, config)
-      {email, response}
-    end
+  Same as `deliver_later!/2` but does not return an ok tuple and raises on
+  errors.
+
+  If successful, this function only returns the `Email` struct.
+
+  If the email is invalid, this function raises an error.
+  """
+  def deliver_later!(_email, _opts \\ []) do
+    raise @cannot_call_directly_error
   end
 
   @doc false
-  def deliver_now(adapter, email, config, _opts) do
-    email = email |> validate_and_normalize(adapter)
+  def deliver_now(adapter, email, config, opts) do
+    with {:ok, email} <- validate_and_normalize(email, adapter) do
+      if empty_recipients?(email) do
+        debug_unsent(email)
+
+        {:ok, email}
+      else
+        debug_sent(email, adapter)
+
+        case adapter.deliver(email, config) do
+          {:ok, response} -> format_response(email, response, opts)
+          {:error, _} = error -> error
+        end
+      end
+    end
+  end
 
-    if email.to == [] && email.cc == [] && email.bcc == [] do
-      debug_unsent(email)
+  defp format_response(email, response, opts) do
+    put_response = Keyword.get(opts, :response, false)
+
+    if put_response do
+      {:ok, email, response}
     else
-      debug_sent(email, adapter)
-      adapter.deliver(email, config)
+      {:ok, email}
     end
+  end
 
-    email
+  @doc false
+  def deliver_now!(adapter, email, config, opts) do
+    case deliver_now(adapter, email, config, opts) do
+      {:ok, email, response} -> {email, response}
+      {:ok, email} -> email
+      {:error, error} -> raise error
+    end
   end
 
   @doc false
   def deliver_later(adapter, email, config) do
-    email = email |> validate_and_normalize(adapter)
+    with {:ok, email} <- validate_and_normalize(email, adapter) do
+      if empty_recipients?(email) do
+        debug_unsent(email)
+      else
+        debug_sent(email, adapter)
+        config.deliver_later_strategy.deliver_later(adapter, email, config)
+      end
 
-    if email.to == [] && email.cc == [] && email.bcc == [] do
-      debug_unsent(email)
-    else
-      debug_sent(email, adapter)
-      config.deliver_later_strategy.deliver_later(adapter, email, config)
+      {:ok, email}
+    end
+  end
+
+  @doc false
+  def deliver_later!(adapter, email, config) do
```

Comment:
I like how the functions with the bangs build on their plain counterparts here. Makes what is changing in this PR clear, too.